tla/raft_spec/README.md

```diff
@@ -0,0 +1,39 @@
+# TLA+ model of CCF's implementation of Raft
+
+The given model consists of three files:
+
+- `ccfraft.tla` : The core specification that implements the features as seen below.
+- `MCraft.tla` : The model checking implementation for the spec. Sets the constants and can be modified for each run to
+  fine tune the settings.
+- `MCraft.cfg` : The core configuration which invariants are to be checked etc. Usually stays untouched during normal
+  model checking.
+
+To run this spec, run the `MCraft.tla` with the `-deadlock` parameter to disable liveness checking (see explanation
+below): `tlc MCraft.tla -deadlock`. See the Readme in the parent folder for a full explanation on running the model checker.
+
+## Implemented features
+
+| Feature | Description |
+| :------ | :---------- |
+| Signatures | In CCF, only entries signed by the primary can be committed (in contrast to all entries in normal Raft) |
+| Reconfiguration | CCF implements a one-transaction version of reconfiguration for Raft (read the description [here](https://microsoft.github.io/CCF/main/overview/consensus.html)). |
+
+## Checked invariants
+
+| Name | Short description |
+| :--- | :---------------- |
+| LogInv | Committing an entry never changes the previous log indices, neither does it decrease the size of the committed log. |
+| CandidateTermNotInLogInv | Candidates that have a chance to win an election will not be elected to a term that already had a leader that appended entries. |
```

Comment:
This sounds a little weaker than it could be. Why do we need the "that appended entries" condition? I believe we should only ever have a single leader in each term - is that not true under our current reconfiguration approach?

Reply:
It isn't, because members of an old config that haven't observed commit of the reconfiguration can elect a leader, even after the new (sufficiently non-overlapping) config has elected its own leader.

It is fine for that to happen, because the leader elected in the old config will only be able to "catch it up" to the reconfiguration, at which point its members will retire. So there may be two leaders at the same term, but only one of them would ever eventually be able to perform writes. The other is only re-transmitting in a dead-end config subset.